ompi/mca/osc/base/osc_base_init.c

```diff
@@ -74,6 +74,109 @@ ompi_osc_base_select(ompi_win_t *win,
     opal_output_verbose( 10, ompi_osc_base_framework.framework_output,
                          "select: component %s selected",
                          best_component->osc_version.mca_component_name );
-
+    
     return best_component->osc_select(win, base, size, disp_unit, comm, info, flavor, model);
+    
+}
+
+int
+ompi_osc_base_iselect(ompi_win_t *win,
+                     void **base,
+                     size_t size,
+                     int disp_unit,
+                     ompi_communicator_t *comm,
+                     ompi_communicator_t **newcomm,
+                     opal_info_t *info,
+                     int flavor,
+                     ompi_request_t **request,
+                     int *model)
+{
+    opal_list_item_t *item;
+    ompi_osc_base_component_t *best_component = NULL;
+    int best_priority = -1, priority;
+
+    if (opal_list_get_size(&ompi_osc_base_framework.framework_components) <= 0) {
+        /* we don't have any components to support us... */
+        return OMPI_ERR_NOT_SUPPORTED;
+    }
+
+    for (item = opal_list_get_first(&ompi_osc_base_framework.framework_components) ;
+         item != opal_list_get_end(&ompi_osc_base_framework.framework_components) ;
+         item = opal_list_get_next(item)) {
+        ompi_osc_base_component_t *component = (ompi_osc_base_component_t*)
+            ((mca_base_component_list_item_t*) item)->cli_component;
+
+        priority = component->osc_query(win, base, size, disp_unit, comm, info, flavor);
+        if (priority < 0) {
+            if (MPI_WIN_FLAVOR_SHARED == flavor && OMPI_ERR_RMA_SHARED == priority) {
+                /* NTH: quick fix to return OMPI_ERR_RMA_SHARED */
+                return OMPI_ERR_RMA_SHARED;
+            }
+            continue;
+        }
+
+        if (priority > best_priority) {
+            best_component = component;
+            best_priority = priority;
+        }
+    }
+
+    if (NULL == best_component) return OMPI_ERR_NOT_SUPPORTED;
+
+    opal_output_verbose( 10, ompi_osc_base_framework.framework_output,
+                         "select: component %s selected",
+                         best_component->osc_version.mca_component_name );
+    
+    return best_component->osc_iselect(win, base, size, disp_unit, comm, newcomm, info, flavor, 
+                                       request, model);
+    
+}
+
+   
+int ompi_osc_complete_base_iselect(ompi_win_t *win,
```

Comment:
What calls this function? The `MPI_WIN_ICREATE_DYNAMIC(..., &request)` will be completed by the user calling `MPI_WAIT(&request)`. How does that figure out what it needs to do?

The selection of which component to use to support this window has already been made (in the `ompi_osc_base_iselect` function), so it should not be selected a second time here.

The only thing that is needed is to call the `osc_complete_iselect` function on the right component. The request should probably be used to store a function pointer to that function - the progress engine (activated by `MPI_WAIT`) can then look at the request, find the function pointer and call it.

That makes this function unnecessary, I think. We just need to do `request.action_function = &best_component->osc_complete_iselect` and `request.action_args = my_icreate_dynamic_args_struct` (or similar) in the `ompi_osc_base_iselect` function.

Reply:
```
 ompi_request_complete_fn_t req_complete_cb; 
 void *req_complete_cb_data;
```
you are right, I was not aware of this... Not sure how to do it just now, but this is a much better option.